src/sentry/api/validators/project_codeowners.py

```diff
@@ -53,40 +43,44 @@ def validate_codeowners_associations(
         external_name__in=usernames + team_names,
         organization_id=project.organization_id,
         provider__in=[ExternalProviders.GITHUB.value, ExternalProviders.GITLAB.value],
-    ).select_related("actor")
+    )
 
     # Convert CODEOWNERS into IssueOwner syntax
     users_dict = {}
     teams_dict = {}
     teams_without_access = []
     users_without_access = []
-    for external_actor in external_actors:
-        type = actor_type_to_string(external_actor.actor.type)
-        if type == "user":
-            try:
-                user: RpcUser = external_actor.actor.resolve()
-            except User.DoesNotExist:
-                continue
-            organization_members_ids = OrganizationMember.objects.filter(
-                user_id=user.id, organization_id=project.organization_id
-            ).values_list("id", flat=True)
-            team_ids = OrganizationMemberTeam.objects.filter(
-                organizationmember_id__in=Subquery(organization_members_ids)
-            ).values_list("team_id", flat=True)
-            projects = Project.objects.get_for_team_ids(Subquery(team_ids))
-
-            if project in projects:
-                users_dict[external_actor.external_name] = user.email
-            else:
-                users_without_access.append(f"{user.get_display_name()}")
-        elif type == "team":
-            team = external_actor.actor.resolve()
-            # make sure the sentry team has access to the project
-            # tied to the codeowner
-            if project in team.get_projects():
-                teams_dict[external_actor.external_name] = f"#{team.slug}"
-            else:
-                teams_without_access.append(f"#{team.slug}")
+
+    team_ids_to_external_names: Mapping[int, str] = {
+        xa.team_id: xa.external_name for xa in external_actors if xa.team_id is not None
+    }
+    user_ids_to_external_names: Mapping[int, str] = {
+        xa.user_id: xa.external_name for xa in external_actors if xa.user_id is not None
+    }
+
+    for user in user_service.get_many(
+        filter=dict(user_ids=list(user_ids_to_external_names.keys()))
+    ):
+        organization_members_ids = OrganizationMember.objects.filter(
+            user_id=user.id, organization_id=project.organization_id
+        ).values_list("id", flat=True)
+        team_ids = OrganizationMemberTeam.objects.filter(
+            organizationmember_id__in=Subquery(organization_members_ids)
+        ).values_list("team_id", flat=True)
+        projects = Project.objects.get_for_team_ids(Subquery(team_ids))
```

Comment:
You'll make 3N queries here. Should we bulk load members/teams/projects?

Reply:
You are correct, I was surprised because this is unfortunately... how it was written before.  I've actually done some improvement because if you check above, we used to loop and call Actor.resolve() for every single entry.

I can take a closer look at more optimization.

Reply:
@markstory Considering this logic already existed before, would it be reasonable for me to leave it?  The problem is simply that making this more efficient results in a bunch of careful near querying to replace it -- we don't really have the tools to granular make connections between users to projects in bulk helper methods.  Which means that if I make the change you're asking, I run the risk of inflicting more bugs for HC to hunt down.  The approach I did take DID reduce some bulk querying (the other user resolve lookup) but leaves the local database inefficiency "as is" to preserve logic.  Is it fair to leave this problem to the owning team?

Reply:
FWIW I did take a stab but realized that it is more complex than first appears.

Reply:
Probably best to leave it alone then. It has worked this way for a while now.